python/lsst/sims/catUtils/mixins/VariabilityMixin.py

```diff
@@ -238,32 +252,75 @@ def applyAgn(self, params, expmjd_in):
         sfint['z'] = params['agn_sfz']
         sfint['y'] = params['agn_sfy']
         tau = params['agn_tau']
-        epochs = expmjd - toff
-        if epochs.min() < 0:
+
+        # A string made up of this AGNs variability parameters that ought
+        # to uniquely identify it.
+        #
+        cache_name = '%d_%.12f_%.12f_%.12f_%.12f_%.12f_%.12f_%.12f_%.12f' \
+        %(seed, sfint['u'], sfint['g'], sfint['r'], sfint['i'], sfint['z'],
+          sfint['y'], tau, toff)
+
+        resumption = False
+
+        # Check to see if this AGN has already been simulated.
+        # If it has, see if the previously simulated MJD is
+        # earlier than the first requested MJD.  If so,
+        # use that previous simulation as the starting point.
+        #
+        if cache_name in _agn_lc_cache:
+            if _agn_lc_cache[cache_name]['mjd'] <expmjd:
+                resumption = True
+
+        if resumption:
+            rng = copy.deepcopy(_agn_lc_cache[cache_name]['rng'])
+            start_date = _agn_lc_cache[cache_name]['mjd']
+            dx_0 = _agn_lc_cache[cache_name]['dx']
+        else:
+            start_date = toff
+            rng = numpy.random.RandomState(seed)
+            dx_0 = {}
+            for k in sfint:
+                dx_0[k]=0.0
+
+        endepoch = expmjd - start_date
+
+        if endepoch < 0:
             raise RuntimeError("WARNING: Time offset greater than minimum epoch.  " +
                                "Not applying variability. "+
                                "expmjd: %e should be > toff: %e  " % (expmjd, toff) +
                                "in applyAgn variability method")
 
-        endepoch = epochs.max()
-
         dt = tau/100.
         nbins = int(math.ceil(endepoch/dt))
-        dt = (endepoch/nbins)/tau
-        sdt = math.sqrt(dt)
-        numpy.random.seed(seed=seed)
-        es = numpy.random.normal(0., 1., nbins)
+
+        x1 = (nbins-1)*dt
+        x2 = (nbins)*dt
+
+        dt = dt/tau
+        es = rng.normal(0., 1., nbins)*math.sqrt(dt)
+        dx_cached = {}
+
         for k in sfint.keys():
-            dx = numpy.zeros(nbins+1)
-            dx[0] = 0.
+            dx2 = dx_0[k]
             for i in range(nbins):
                 #The second term differs from Zeljko's equation by sqrt(2.)
                 #because he assumes stdev = sfint/sqrt(2)
-                dx[i+1] = -dx[i]*dt + sfint[k]*es[i]*sdt + dx[i]
-            x = numpy.linspace(0, endepoch, nbins+1)
-            intdx = interp1d(x, dx)
-            magoff = intdx(epochs)
-            dMags[k] = magoff
+                dx1 = dx2
+                dx2 = -dx1*dt + sfint[k]*es[i] + dx1
+
+            dx_cached[k] = dx2
+            dMags[k] = (endepoch*(dx1-dx2)+dx2*x1-dx1*x2)/(x1-x2)
+
```

Comment:
Add a comment on why the cache is being reset. Is this physical, or a space saving device.